plutus-core/plutus-ir/src/PlutusIR/Core.hs

```diff
@@ -1,5 +1,35 @@
-module PlutusIR.Core (module Export) where
+module PlutusIR.Core (
+    module Export,
+    Expr (..),
+    collectArgs,
+    mkTermApps,
+) where
+
+import Data.List (foldl')
 
 import PlutusIR.Core.Instance ()
 import PlutusIR.Core.Plated as Export
 import PlutusIR.Core.Type as Export
+
+data Expr tyname name uni fun ann
```

Comment:
I don't love this naming. I would expect "expression" to be synonymous with "term", this isn't that. What was wrong with "arg"? It _is_ representing arguments?

Reply:
It's just "either Term or Type", which can be arguments, but doesn't have to be. And I don't like the name `TermOrType`.

GHC's `Expr` can be either term or type, and this just follows that naming.

Reply:
It's a datatype for a _specific use_, namely representing arguments. What's wrong with that?

> GHC's Expr can be either term or type, and this just follows that naming.

I don't understand this argument. GHC's `Expr` is our `Term`. Adding something that isn't the main term type called "Expr" is just confusing IMO.

Reply:
> It's a datatype for a specific use, namely representing arguments. What's wrong with that?

Nothing terribly wrong, but in such cases my preference is to name the datatype properly, and use a synonym `type Arg = <proper name>`

> I don't understand this argument. GHC's Expr is our Term. Adding something that isn't the main term type called "Expr" is just confusing IMO.

My point is GHC's `Expr` has a `Type` constructor so it can be either a term or a type. But if "expression" and "term" are often used interchangeably in the literature then that is indeed not a good name.

Reply:
> Nothing terribly wrong, but in such cases my preference is to name the datatype properly, and use a synonym type Arg = <proper name>

Well, it's not clear what the "proper" name for this is? It's a datatype for either a term or a type. The only place that really comes up is a genericised version of "arguments"; so it seems like the sensible names are either `Arg` or `TermOrType`.

> But if "expression" and "term" are often used interchangeably in the literature then that is indeed not a good name.

Seems like they're pretty interchangeable, or used ambiguously: https://cs.stackexchange.com/a/64702